### PR TITLE
[None][perf] default NumExpr to one thread and lazy-load it

### PR DESCRIPTION
**Pull request by dhansen-nvidia**
Default NumExpr to one thread and keep it out of the normal inference import path unless a skip-softmax checkpoint formula actually needs it.

## Why this exists

This investigation started with an InferenceX comparison where the CI-equivalent `im-gb300` run produced P90 interactivity of 79.39 tok/s/user and p50 TTFT of 12.54 s, versus 91.22 tok/s/user and 7.27 s on AGA. Profiling showed about 420 mostly idle native math-library threads per decode rank contending with the engine step. Setting `OMP_NUM_THREADS=1` on `im-gb300` restored P90 interactivity to 91.30 tok/s/user and p50 TTFT to 7.68 s.

That established that host thread contention mattered, but not that NumExpr alone caused the performance difference. The subsequent TRT-LLM audit found that the normal attention import path loaded NumExpr in each process even though TRT-LLM uses it only to validate and evaluate scalar skip-softmax formulas. NumExpr creates a process-global worker pool when it imports with more than one thread, so the process-per-rank execution model multiplies threads which provide no useful parallelism for this work.

## With these changes

- TRT-LLM defaults `NUMEXPR_NUM_THREADS` to `1` during package bootstrap, before it imports PyTorch or NumExpr. An explicit application setting is preserved.
- `SkipSoftmaxFormula` imports NumExpr only when a checkpoint formula is actually consumed. The existing NumExpr expression contract and dependency remain unchanged.
- The default is intentionally not `NUMEXPR_MAX_THREADS=1`. If a future feature has enough NumExpr work to benefit from parallel evaluation, it can opt into a larger process-global pool explicitly.
- The documentation records the import-order limitation: applications that import NumExpr before TRT-LLM must configure it before process startup.

With one thread selected before NumExpr is imported, its native initialization path does not create helper threads; the scalar formulas execute on the calling thread.

## What this does not establish

The OMP-only run and the combined OpenBLAS/NumExpr/OMP run were equivalent within run noise. The Enroot PyTorch hook may also explain part of the cluster difference because it injects `OMP_NUM_THREADS=1` for multi-task Slurm jobs when the value is otherwise unset. I would not treat this PR by itself as proof that NumExpr caused the 13–15% regression; it fixes the eager import and unnecessary per-process NumExpr pools exposed by that investigation.

## Validation

Pre-commit, Python compilation, the bootstrap default/override check, and `git diff --check` pass. The new fresh-process tests cover the lazy import, the one-thread default, and an explicit two-thread override. I could not run those tests locally because this host does not have PyTorch, pytest, or NumExpr installed; they remain pending CI.

The LLM-argument golden manifest is unchanged because this PR does not add or modify an LLM argument.

## Dev Engineer Review

- `_bootstrap.py` now defaults `NUMEXPR_NUM_THREADS` to `1` without overriding explicit settings.
- `SkipSoftmaxFormula` now imports NumExpr only when formula validation or threshold computation runs.
- The NumExpr expression contract and dependency remain unchanged.
- Documentation states that applications must configure NumExpr before importing TRT-LLM when they need custom settings.
- No public API changes or manifest changes were identified.

## QA Engineer Review

- Modified `tests/unittest/_torch/attention/sparse/test_skip_softmax_formula.py`.
- Tests now cover function-local NumExpr imports, the default one-thread configuration, explicit thread configuration, and thread counts before and after formula evaluation.
- Subprocess-based checks cover fresh-process environment behavior.
- Local fresh-process execution remains pending because PyTorch, pytest, and NumExpr are unavailable.
- No matching CI or manual-QA test-list entry was identified.
- Coverage verdict: needs follow-up.

## Per-File QA Perspective

- `docs/source/features/sparse-attention.md`: Documents lazy NumExpr loading, the default thread setting, and explicit application configuration requirements.
- `docs/source/visual-gen/features/sparse-attention.md`: Documents the same NumExpr behavior for the visual-generation feature documentation.
- `tensorrt_llm/_bootstrap.py`: Adds an environment default before remaining bootstrap setup. Verify that it runs before PyTorch or NumExpr imports and preserves explicit values.
- `tensorrt_llm/_torch/attention/backends/sparse/skip_softmax/params.py`: Removes the module-level NumExpr import and adds function-local imports. Verify formula validation and threshold computation, including missing or invalid checkpoint formulas.
- `tests/unittest/_torch/attention/sparse/test_skip_softmax_formula.py`: Covers lazy imports and default or explicit NumExpr thread settings through subprocess tests. The tests are not listed in the inspected CI or manual-QA test lists.

### CODE REVIEW

**Review by BowenFu — APPROVED** [review]

**Review by sylvesterkaczmarek — COMMENTED** [review]
I checked the import-order and ownership assumptions behind this change rather than only the formula code. `tensorrt_llm.__init__` imports only the stdlib-only bootstrap first and calls `_prepare_environment()` before importing `torch`, so the `NUMEXPR_NUM_THREADS` default is established before either Torch or the later sparse-attention path can indirectly initialize NumExpr. Using `os.environ.setdefault()` also preserves an application's explicit value rather than overwriting it.

On the consumer side, removing the module-level `numexpr` import keeps the normal attention import path clean; `_load_numexpr()` is reached only when `SkipSoftmaxFormula` is actually validated or evaluated. The fresh-process tests cover both the lazy-import boundary and an explicit two-thread override, which are the two cases most likely to regress this behavior.

I also checked the current head's GitHub workflows: PR Checks, Release Checks, and LLM API Compatibility are green. I don't see a blocking lifecycle or compatibility issue in the current change.

File: tensorrt_llm/_torch/attention_backend/sparse/skip_softmax/params.py [hunk 1]
```diff
@@ -126,7 +132,8 @@ class SkipSoftmaxFormula(StrictBaseModel):
     )
 
     @model_validator(mode="after")
-    def _validate_formula(self):
+    def _validate_formula(self) -> "SkipSoftmaxFormula":
+        numexpr = _load_numexpr()
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/sparse/skip_softmax/params.py]: Can we use local import instead of a new helper function?

Comment (dhansen-nvidia) [reply tensorrt_llm/_torch/attention_backend/sparse/skip_softmax/params.py]: Yes, this is cleaner. I switched both call sites to direct local imports, which preserves lazy initialization without the helper.

**Review by BowenFu — APPROVED** [review]

**Review by yuxianq — APPROVED** [review]

**Review by sylvesterkaczmarek — COMMENTED** [review]
Re-reviewed current `cdac291a`. The bootstrap invariant remains correct: `_prepare_environment()` calls `_set_numexpr_thread_default()` before package imports can load NumExpr, and `setdefault("NUMEXPR_NUM_THREADS", "1")` preserves an explicit user override. The skip-softmax code now imports NumExpr only inside the formula methods, and the subprocess tests verify both the default and configured effective thread counts. I do not see a remaining code blocker from the thread-initialization concern I raised. NVIDIA's current internal L0 run is still failing, so that CI result should be resolved separately before merge.

**Review by BowenFu — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]

**Review by juney-nvidia — APPROVED** [review]
Approved from doc perspective.

File: tests/unittest/_torch/attention/sparse/test_skip_softmax_formula.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def _run_python(code: str, numexpr_threads: str | None = None) -> subprocess.CompletedProcess[str]:
+    env = os.environ.copy()
+    env.pop("NUMEXPR_MAX_THREADS", None)
+    env.pop("NUMEXPR_NUM_THREADS", None)
+    if numexpr_threads is not None:
+        env["NUMEXPR_NUM_THREADS"] = numexpr_threads
+
+    return subprocess.run(
```
[the remark this answers is not in the document]
Comment (dhansen-nvidia) [reply tests/unittest/_torch/attention/sparse/test_skip_softmax_formula.py:21]: Fixed in d1af1c1f52. The child process now has a 300-second timeout. The probe took about 17 seconds per case on ARM in CI, so I used a larger bound than 30 seconds to avoid making the test brittle under load while still preventing an unbounded hang.
